workload/schemachange: FK validation does not check inserted rows

fqazi opened the pull request:
Fixes: #63064

Previously, the validation of foreign keys during inserts,
did not take into account self referential scenarios properly,
where rows inside the insert would allow the FK constraint to
be met. To address this, this patch updates the validation to
scan the set of inserted rows for self referential foreign keys.

Release note: None

fqazi committed:
workload/schemachange: FK validation does not check inserted rows

Fixes: #63064

Previously, the validation of foreign keys during inserts,
did not take into account self referential scenarios properly,
where rows inside the insert would allow the FK constraint to
be met. To address this, this patch updates the validation to
scan the set of inserted rows for self referential foreign keys.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82866)
<!-- Reviewable:end -->

fqazi committed:
workload/schemachanger: unique constraint validation is buggy for exprs

Previously, when we were validating unique constraint expressions during
inserts we did not properly handle NULL values. This could lead to us
incorrectly skipping over evaluation trying to determine the uniqueness
of values. This could mean both valid / invalid inserts were not properly
fully evaluated leading to unexpected errors. To address this, this
patch properly counts the number of NULL values in the unique expression
where if all of them are NULL then the expression is not evaluated for
uniqueness.

Release note: None

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82866#-:-N4ZNNAgB2-7VMmCwUu6:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[`pkg/workload/schemachange/error_screening.go` line 1149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82866#-N4ZMVY800eklYkOyJNM:-N4ZMVY99SY0hfseuF_s:b-f1t27m) ([raw file](https://github.com/cockroachdb/cockroach/blob/63aabd940fbea83619b760352b07a1f8d8dbe561/pkg/workload/schemachange/error_screening.go#L1149)):*
> ```Go
> 			}
> 			matches, err := og.scanBool(ctx, tx,
> 				fmt.Sprintf("SELECT %s = %s", parentValueInSameInsert, childValue))
> ```

you could turn this into a single query with an array: `SELECT true =  ANY (ARRAY[1 = 1, 2 = 1]);` furthermore, you could put it into the same query you're going to run below. 



_Code quote:_
```Go
			matches, err := og.scanBool(ctx, tx,
				fmt.Sprintf("SELECT %s = %s", parentValueInSameInsert, childValue))
```


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
@ajwerner TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82866#-:-N4_TK15974SGhoOqbND:b-gz014k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

fqazi commented:
bors r+
